pandas/core/frame.py

```diff
@@ -10482,38 +10483,82 @@ def _get_data() -> DataFrame:
-                result = df._reduce(
-                    op,
-                    name,
-                    axis=0,
-                    skipna=skipna,
-                    numeric_only=False,
-                    filter_type=filter_type,
-                    **kwds,
-                ).iloc[:0]
-                result.index = df.index
-                return result
-            df = df.T
-
-        # After possibly _get_data and transposing, we are now in the
-        #  simple case where we can use BlockManager.reduce
-        res = df._mgr.reduce(blk_func)
-        out = df._constructor(res).iloc[0]
-        if out_dtype is not None:
-            out = out.astype(out_dtype)
-        elif (df._mgr.get_dtypes() == object).any():
-            out = out.astype(object)
-        elif len(self) == 0 and name in ("sum", "prod"):
-            # Even if we are object dtype, follow numpy and return
-            #  float64, see test_apply_funcs_over_empty
-            out = out.astype(np.float64)
 
-        return out
+        axis = self._get_axis_number(axis)
+        assert axis in [0, 1]
+
+        if len(df._mgr) > 0:
+            common_dtype = find_common_type(list(df._mgr.get_dtypes()))
+            is_masked_ea = isinstance(common_dtype, BaseMaskedDtype)
+            is_np = isinstance(common_dtype, np.dtype)
+        else:
+            common_dtype = None
+
+        if axis == 0 or common_dtype is None or not (is_masked_ea or is_np):
+            if axis == 1:
+                if len(df.index) == 0:
+                    # Taking a transpose would result in no columns, losing the dtype.
+                    # In the empty case, reducing along axis 0 or 1 gives the same
+                    # result dtype, so reduce with axis=0 and ignore values
+                    result = df._reduce(
+                        op,
+                        name,
+                        axis=0,
+                        skipna=skipna,
+                        numeric_only=False,
+                        filter_type=filter_type,
+                        **kwds,
+                    ).iloc[:0]
+                    result.index = df.index
+                    return result
+                df = df.T
+
+            # After possibly _get_data and transposing, we are now in the
+            #  simple case where we can use BlockManager.reduce
+            res = df._mgr.reduce(blk_func)
+            out = df._constructor(res).iloc[0]
+            if out_dtype is not None:
+                out = out.astype(out_dtype)
+            elif (df._mgr.get_dtypes() == object).any():
+                out = out.astype(object)
+            elif len(self) == 0 and name in ["sum", "prod"]:
+                # Even if we are object dtype, follow numpy and return
+                #  float64, see test_apply_funcs_over_empty
+                out = out.astype(np.float64)
+
+            return out
+
+        if is_np or not is_numeric_dtype(common_dtype):
+            values = df.values
```

Comment:
not really relevant here but in general df._values is better than df.values, preserves e.g. DatetimeArray[tzaware]